### PR TITLE
Dont call multiunzip when no stats

### LINKED ISSUE #9219: Improve performance of `list_files_for_scan` when not collecting statistics
### Is your feature request related to a problem or challenge?

Right now DataFusions planning performance is a bottleneck for our application.  We noticed that there is a non-negligible amount of work being done in `get_statistics_with_limit` even when collecting statistics is disabled.  In particular the work done within `multiunzip` which constructs the `ColumnStatistics` is what we view as being unnecessary when collecting statistics is disabled.  We would like to add some logic to improve performance when collect statistics is disabled.

### Describe the solution you'd like

On the `list_files_for_scan` method of `ListingTable` we would like to update the logic for getting the list of `PartitionedFiles` based on the value of `self.options.collect_stat`.

So going from
```rust
let (files, statistics) = get_statistics_with_limit(files, self.schema(), limit).await?;
```

to something like
```rust
let (files, statistics) = match self.options.collect_stat {
    true => get_statistics_with_limit(files, self.schema(), limit).await?,
    false => get_files_with_unknown_stats(files, self.schema(), limit).await?
}
```

Where `get_files_with_unknown_stats` avoids the call to `multiunzup`.

### Describe alternatives you've considered

An alternative approach could be adding a parameter to `get_statistics_with_limit` for `collect_stats` and calling `multiunzip` based on that.

### Additional context

Our application has low latency requirement and in our current setup DataFusion's planning performance is our bottleneck.  We will eventually be turning on statistics collection but right now we cant and so we are looking to improve planning performance where we can.

Based on our internal benchmarks we saw physical planning performance improve by 16-43% after making the above mentioned change (we have a lot of files so impact can be large, for smaller number of files the impact will probably not be as big or could be neglible).

### PR DESCRIPTION
**Pull request by matthewmturner**
## Which issue does this PR close?

Closes #9219

### CODE REVIEW

File: datafusion/core/benches/sql_planner.rs [hunk 1]
```diff
@@ -234,7 +234,7 @@ fn criterion_benchmark(c: &mut Criterion) {
         let sql = std::fs::read_to_string(format!("../../benchmarks/queries/{}.sql", q))
             .unwrap();
         c.bench_function(&format!("physical_plan_tpch_{}", q), |b| {
-            b.iter(|| logical_plan(&ctx, &sql))
+            b.iter(|| physical_plan(&ctx, &sql))
```
Comment (matthewmturner) [inline datafusion/core/benches/sql_planner.rs:237]: There was disconnect between benchmark name (physical plan) and the function(logical plan), so I corrected this

**Review by Dandandan — APPROVED** [review]

File: datafusion/core/src/datasource/statistics.rs [hunk 2]
```diff
@@ -35,6 +35,7 @@ pub async fn get_statistics_with_limit(
     all_files: impl Stream<Item = Result<(PartitionedFile, Statistics)>>,
     file_schema: SchemaRef,
     limit: Option<usize>,
+    collect_stats: bool,
```
Comment (alamb) [inline datafusion/core/src/datasource/statistics.rs:38]: Given the function is called 'get_statistics` and now there is a boolean flag that implies the stats aren't collected, can we please add some documentation to the docstring explaining what the `collect_stats` flag does?

**Review by alamb — APPROVED** [review]
Thank you @matthewmturner  and @Dandandan

File: datafusion/core/src/datasource/statistics.rs (hunk 2 above)
Comment (matthewmturner) [reply datafusion/core/src/datasource/statistics.rs:38]: Yes good catch, will do.

**Comment by matthewmturner** [comment]
@alamb I have updated the docstring. let me know if anything else

**Comment by alamb** [comment]
Thanks again @matthewmturner !

**Pull request merged**